wkcuber/api/Layer.py

```diff
@@ -99,20 +118,157 @@ def set_bounding_box_size(self, size: Tuple[int, int, int]) -> None:
+                    right_offset[i] = current_offset[i] + current_size[i]
+                    mag.write(
+                        data=np.zeros(padding_shape, dtype=mag.get_dtype()),
+                        offset=right_offset,
+                    )
+
+        parsed_interpolation_mode = parse_interpolation_mode(
+            interpolation_mode, self.name
+        )
+        prev_mag = from_mag
+        target_mag = get_next_mag(prev_mag, scale)
+
+        while target_mag <= max_mag:
+            assert prev_mag < target_mag
+            assert target_mag.to_layer_name() not in self.mags
+
+            prev_mag_ds = self.mags[prev_mag.to_layer_name()]
+
+            mag_factors = [
+                t // s for (t, s) in zip(target_mag.to_array(), prev_mag.to_array())
+            ]
+
+            # initialize the new mag
+            target_mag_ds = self._initialize_mag_from_other_mag(
+                target_mag, prev_mag_ds, compress
+            )
+
+            # Get target view
+            target_mag_view = target_mag_ds.get_view(is_bounded=False)
+
+            # perform downsampling
+            with get_executor_for_args(args) as executor:
+                voxel_count_per_cube = np.prod(prev_mag_ds._get_file_dimensions())
+                job_count_per_log = math.ceil(
+                    1024 ** 3 / voxel_count_per_cube
+                )  # log every gigavoxel of processed data
+
+                if buffer_edge_len is None:
+                    buffer_edge_len = determine_buffer_edge_len(
+                        prev_mag_ds.view
+                    )  # DEFAULT_EDGE_LEN
+                job_args = (
+                    mag_factors,
+                    parsed_interpolation_mode,
+                    buffer_edge_len,
+                    compress,
+                    job_count_per_log,
+                )
+                prev_mag_ds.get_view().for_zipped_chunks(
+                    # this view is restricted to the bounding box specified in the properties
+                    downsample_cube_job,
+                    target_view=target_mag_view,
+                    job_args_per_chunk=job_args,
+                    source_chunk_size=np.array(target_mag_ds._get_file_dimensions())
+                    * mag_factors,
+                    target_chunk_size=target_mag_ds._get_file_dimensions(),
+                    executor=executor,
+                )
+
+            logging.info("Mag {0} successfully cubed".format(target_mag))
+
+            prev_mag = target_mag
+            target_mag = get_next_mag(target_mag, scale)
+
+    def setup_mag(self, mag: Union[str, Mag]) -> None:
         pass
 
 
 class WKLayer(Layer):
     mags: Dict[str, WKMagDataset]
 
-    def add_mag(
-        self,
-        mag: Union[str, Mag],
-        block_len: int = None,
-        file_len: int = None,
-        block_type: int = None,
-    ) -> MagDataset:
+    def add_mag(self, mag: Union[str, Mag], **kwargs: Any) -> MagDataset:
```

Comment:
Unfortunately, I had to change the signature of the method to add this method to the base class. This is necessary to pass the typecheck. I do not understand why this was not a problem in #249